coderd/notifications/notifier.go

```diff
@@ -0,0 +1,247 @@
+		case <-n.quit:
+			return nil
+		case <-n.tick.C:
+			// sleep until next invocation
+		}
+	}
+}
+
+// process is responsible for coordinating the retrieval, processing, and delivery of messages.
+// Messages are dispatched concurrently, but they may block when success/failure channels are full.
+//
+// NOTE: it is _possible_ that these goroutines could block for long enough to exceed CODER_NOTIFICATIONS_DISPATCH_TIMEOUT,
+// resulting in a failed attempt for each notification when their contexts are canceled; this is not possible with the
+// default configurations but could be brought about by an operator tuning things incorrectly.
+func (n *notifier) process(ctx context.Context, success chan<- dispatchResult, failure chan<- dispatchResult) error {
+	n.log.Debug(ctx, "attempting to dequeue messages")
+
+	msgs, err := n.fetch(ctx)
+	if err != nil {
+		return xerrors.Errorf("fetch messages: %w", err)
+	}
+
+	n.log.Debug(ctx, "dequeued messages", slog.F("count", len(msgs)))
+	if len(msgs) == 0 {
+		return nil
+	}
+
+	var eg errgroup.Group
+	for _, msg := range msgs {
+		// A message failing to be prepared correctly should not affect other messages.
+		deliverFn, err := n.prepare(ctx, msg)
+		if err != nil {
+			n.log.Warn(ctx, "dispatcher construction failed", slog.F("msg_id", msg.ID), slog.Error(err))
+			failure <- newFailedDispatch(n.id, msg.ID, err, false)
+			continue
+		}
+
+		eg.Go(func() error {
+			// Dispatch must only return an error for exceptional cases, NOT for failed messages.
+			return n.deliver(ctx, msg, deliverFn, success, failure)
+		})
+	}
+
+	if err = eg.Wait(); err != nil {
+		n.log.Debug(ctx, "dispatch failed", slog.Error(err))
+		return xerrors.Errorf("dispatch failed: %w", err)
+	}
+
+	n.log.Debug(ctx, "dispatch completed", slog.F("count", len(msgs)))
+	return nil
+}
+
+// fetch retrieves messages from the queue by "acquiring a lease" whereby this notifier is the exclusive handler of these
+// messages until they are dispatched - or until the lease expires (in exceptional cases).
+func (n *notifier) fetch(ctx context.Context) ([]database.AcquireNotificationMessagesRow, error) {
+	msgs, err := n.store.AcquireNotificationMessages(ctx, database.AcquireNotificationMessagesParams{
+		Count:           int32(n.cfg.LeaseCount),
+		MaxAttemptCount: int32(n.cfg.MaxSendAttempts),
+		NotifierID:      n.id,
+		LeaseSeconds:    int32(n.cfg.LeasePeriod.Value().Seconds()),
+	})
+	if err != nil {
+		return nil, xerrors.Errorf("acquire messages: %w", err)
+	}
+
+	return msgs, nil
+}
+
+// prepare has two roles:
+// 1. render the title & body templates
+// 2. build a dispatcher from the given message, payload, and these templates - to be used for delivering the notification
+func (n *notifier) prepare(ctx context.Context, msg database.AcquireNotificationMessagesRow) (dispatch.DeliveryFunc, error) {
+	select {
+	case <-ctx.Done():
+		return nil, ctx.Err()
+	default:
+	}
```

Comment:
I'm working on REST API to [enable/disable notifiers](https://github.com/coder/team-coconut/issues/4), and I'm curious if we can simplify the logic by removing all "select-case-ctx.Done" logic. In theory, we need the notifier to handle the server shutdown gracefully, so wait until all notification messages are dispatched, right? Does it make a difference to the check the context on every level - fetch/prepare/deliver?

Reply:
I believe we need all of this.

- `fetch` doesn't check the context (but the query does down in `ctxDriverQuery`)
- `prepare` is called inside `process` and the context could be canceled between `fetch` completing and `prepare` being invoked
- `deliver` is the same because it's run in a goroutine which could block for an indeterminate amount of time between `prepare` completing and its being executed

I was imagining that you could simply check for the presence of a lock before `process` is invoked in `run` in each loop and suppress it from executing. What approach were you considering?

Reply:
I want to use `site_configs` to store the notification config, and expose it via API. Then, I will patch process to fetch relevant `site_config`  first and continue dequeuing if enabled.

Reply:
I like the idea of using `site_configs`, but I'd suggest using it as a killswitch instead.
I think this should happen outside of `process`, too, since it's not its responsibility to determine if it can run or not.